src/tensor_parallel/state_actions.py

```diff
@@ -0,0 +1,105 @@
+"""slicing_actions.py: file contains actions """
+
+from abc import ABC, abstractclassmethod
+from typing import Any, Sequence
+
+import torch
+from torch import Tensor
+
+
+class StateAction(ABC):
```

Comment:
What I did with `state_rules`:
* They are now `Dict[str, StateAction]` where `StateAction` is an interface.
* `StateAction` has two methods:
*  * `def __call__(self, tensor: Tensor, rank: int) -> Tensor:` - split the tensor and return the corresponding part.
*  * `def undo(self, tensors: Sequence[Tensor]) -> Tensor:` - merge the parts back for initial tensor reconstruction.